<!DOCTYPE html>
<head>

    <meta charset="utf-8">
    <link rel="stylesheet" href="stylesheets/bootstrap.css">
    <script src="javascripts/jquery-2.2.3.min.js"></script>
    <script src="javascripts/bootstrap.js"></script>
    <script src="https://d3js.org/d3.v3.min.js"></script>

    <script src="https://maps.googleapis.com/maps/api/js?key=AIzaSyDDvbCITGxjpTV2QDBWJQUtd1_vqxwk2mg"></script>

        <style>

        .node {
            cursor: pointer;
        }

        .node circle {
            fill: #fff;
            stroke: red;
            stroke-width: 1.5px;
        }

        .node text {
            font: 10px sans-serif;
        }

        .link {
            fill: none;
            stroke: #ccc;
            stroke-width: 1.5px;
        }
        .wrap{
            word-wrap: break-word;
        }
            #map{
                width:100%;
                height: 400px;
            }

        svg{
            border:1px solid black;
        }

    </style>
</head>
<body class="container">
<div class="row">
    <div class="col-xs-12">
        <h1 class="text-center">MOLECULAR MATINGS</h1>

        <div>
            <!-- Nav tabs -->
            <ul id="myTabs" class="nav nav-tabs" role="tablist">
                <li role="presentation" class="active"><a href="#home" aria-controls="home" role="tab" data-toggle="tab">Home</a></li>
                <li role="presentation"><a href="#about" aria-controls="profile" role="tab" data-toggle="tab">About</a></li>
                <li role="presentation"><a href="#mtdna" aria-controls="messages" role="tab" data-toggle="tab">Mitochondrial DNA</a></li>
            </ul>

            <!-- Tab panes -->
            <div class="tab-content">
                <div role="tabpanel" class="tab-pane active" id="home">
                    <div class="row">
                        <div class="col-xs-8">
                            Haplogroups : <select id="haplogroup"></select>

                            <div id="tree">

                            </div>
                        </div>
                        <div class="col-xs-4">
                            <h3>Family/Haplotype Description</h3>

                            <div class="row">
                                <div class="col-xs-11 wrap">
                                    <p> This Text box would be something that I would edit for each Bruce Lowe family and each mtDNA haplotype. It would be a description of each of the families/haplotypes that would outline where the errors in the Stud Book appeared and where the families join together in sharing the same mitochondrial haplotype.</p>


                                    <p>It would probably be somewhere around 400 words in length at the most and would fill out this column to the bottom of the page. It would change every time that the user changed the lowe family or haplotype that they were looking at.</p>
                                    ​
                                    <p>To the left the family or haplotype would appear. I was hoping it would look somewhat like this http://nicolas.kruchten.com/genealogy/lepage_bandet.html</p>

                                    <p>Each of the mitochondrial haplotypes would need to have a unique colour so that the errors could be clearly seen and visually displayed.</p>
                                    ​
                                    <p>In some cases there could be as many as 20 generations of horses so we need to think about how this is best displayed and allow the user to click on the rendered image to enlarge it or have some way of magnifying branches of interest.</p>

                                </div>

                            </div>
                           </div>

                    </div>
                </div>
                <div role="tabpanel" class="tab-pane" id="about">
                    <div class="row">
                        <div class="col-xs-4">
                            <h3>CONTACT</h3>
                            <div class="row">
                                <div class="col-xs-11 wrap">
                                    <p>I'm a paragraph. Click here to add your own text and edit me. It’s easy. Just click “Edit Text” or double click me to add your own content and make changes to the font. I’m a great place for you to tell a story and let your users know a little more about you.</p>

                                </div>
                                <div class="col-xs-8 col-xs-offset-4">
                                    500 Terry Francois Street<br>
                                    San Francisco, CA 94158<br>
                                    info@mysite.com<br>
                                    Tel: 123-456-7890<br>
                                </div>

                                <div>

                                </div>


                            </div>

                        </div>
                        <div class="col-xs-8">
                               <div id="map"></div>
                        </div>
                    </div>



                </div>
                <div role="tabpanel" class="tab-pane" id="mtdna">


                    <div class="row" >
                        <div class="col-xs-8">

                            <h3>MITOCHONDRIAL DNA</h3>


                            <p> Mitochondrial DNA is a snapshot of the distant matrilineal past. There horses (and other mammals) should resemble the dam of thirty, forty, or more removes about as closely as they do their mother. If comparison of mitochondrial 'photographs' of members of the same TB female family reveals one or more individuals that do not closely resemble the rest it is conclusive proof of error somewhere in the tail female pedigree record. If comparison of two or more different families reveals no difference in appearance it raises the possibility of, but does not prove, a common dam within the last thousand years or so.</p>

                            <p> Mitochondria are the power generators of the cell, present there in multiple copy, the actual number being commensurate with a cell's metabolic activity. The mitochondria have their own DNA in a single ring-shaped chromosome, enabling them to reproduce independently of the cell itself. The mitochondrial chromosome of the domestic horse contains 16,660 nucleotides, over 90% of which control protein synthesis (the "coding region"). The remainder is the displacement loop (d-loop) where the chromosome opens for transcription, also known as the "non-coding" or control region.</p>

                            <p> MtDNA is inherited exlusively from the mother. Any paternal mitochondria that are not expended prior to fertilization are normally inactivated at that time. Lacking a paternal analogue there is effectively no recombination. The only way mtDNA changes is via spontaneous mutations. Such mutations do regularly occur. The average overall rate of mutation in mtDNA is measured in thousands of years. If present in a germ cell mutations persist in subsequent generations. Mutations in the coding region often result in functional problems for the organism. These are less likely to persist than mutations in the d-loop. Within the d-loop are two sections which exhibit a faster than average rate of mutation. These are known as hypervariable sections HVS1 and HVS2.</p>

                            <p>Sets of mutations shared by a large number of individuals define maternally-linked populations known as clades or haplogroups, several of which are large enough to have major subdivisions within them called sub-haplogroups. Within these, additional mutations define minor subdivisions and, finally, individual haplotypes. By comparing shared and non-shared mutations between two or more individuals a rough estimate can be made as to how long ago their most recent common ancestor (MRCA) in tail female lived. This has made mtDNA a very useful tool for studying the evolution of various species. It has also become a popular tool for human genealogical research.</p>

                            <p>It is equally useful for equine genealogical applications such as testing the accuracy of the record of the Thoroughbred female families. To date, four reports correlating TB female families with mtDNA haplotypes have been published. However, sequencing for three of them (Hill et al. 2002, Bower et al. 2012a, and Bower et al. 2012b) was limited to d-loop fragments, usually just partial HVS1. Paralleling the evolution of human mtDNA analysis, d-loop alone was initially regarded by many as sufficient for adequate haplotype definition but turned out to be insufficient for that purpose. As first reported anecdotally by Harrison & Turrion-Gomez 2006, Achilli et al. 2012 proved beyond any doubt that it’s frequently impossible to adequately define equine mtDNA haplotypes without information from the coding region.</p>

                            <p>Since 2012 many full mt TB sequences have been published through GenBank. While, unfortunately, none of them are identified by family number they nonetheless prove that the short partial HVS1 sequences used for the three aforementioned reports effectively masked considerable variation in the TB. What was first reported as a single haplotype, "F", by Hill et al. 2002 (Table 1) and subsequently (by segregation at an HVS1 indel excluded from previous consideration) was reported as two haplotypes by Bower et al. 2012b (Table 1, "Lineages 2, 8, 16" and "Lineages 7, 17, 22"), is actually no fewer than 4 different haplotypes as defined from full mt TB sequences. Bower’s "Lineages 10, 14, 42" (Hill’s "D") probably conflates two, possibly three haplotypes. "Lineages 6, 20, 23" (Hill’s "N") may conflate two haplotypes. Outside their Tables, Bower et al. 2012b effectively conglomerated ("Discussion" p.234) families 1, A1, and A4, conflating no fewer than two different haplotypes. So far, there is no evidence of conflation in the "Lineages 9 and 12" (Hill’s"G") and “Lineages 4, 11, 13” (Hill’s “J”) conglomerates although in HVS1 alone the former is distinguished from several other haplotypes only at mutational ‘hotspots’ which are the last resort in determining haplotype assignment.</p>

                            <p> Sequencing for the fourth TB study (Rogers, 2014) included certain markers in the coding region. The findings reported have clarified the phylogenetic relationships between the families (1, 25, A4) with typical haplotypes in haplogroup N and enabled specific definition of the typical haplotype for family 16, thereby resolving some of the conflation in the "Lineages 2, 8, 16" conglomerate. Full resolution of conflation in that and other conglomerates awaits further study. That is why two or more haplotypes are associated with some families/family branches in the table below which sets forth distribution of 27 haplotypes, presently regarded as representative of matrilineally unrelated foundation era mares, among and within the 31 numbered TB families (27 British, 4 American) sampled and reported publicly to date.</p>

                        </div>

                    </div>


                </div>
            </div>

        </div>
    </div>
</div>
<script>
    $('#myTabs a').click(function (e) {
        e.preventDefault()
        $(this).tab('show')
    })
    function initMap() {
        var mapDiv = document.getElementById('map');
        var map = new google.maps.Map(mapDiv, {
            center: {lat: 38.0406, lng: -84.5037},
            zoom: 8
        });
    }
    initMap();
</script>
<script>

    var dispatch = d3.dispatch("load", "statechange");

    d3.json("/gethaplogroups", function(error, data) {
        data.haplogroup.forEach(function(d){
            d3.select("#haplogroup").append("option").text(d);
        })
        d3.select("#haplogroup").on('change',function(){
            var sel = d3.select('#haplogroup').node()
            var selected =sel.options[sel.selectedIndex].value;
            getHaploGroupData(selected);
        })
    })
</script>
<script>

    var margin = {top: 20, right: 120, bottom: 20, left: 120},
            width = 600 - margin.right - margin.left,
            height = 500 - margin.top - margin.bottom;

    var i = 0,
            duration = 750,
            root;

    var tree = d3.layout.tree()
            .size([height, width]);

    var diagonal = d3.svg.diagonal()
            .projection(function(d) { return [d.y, d.x]; });


    var svg = d3.select("#tree").append("svg")
            .call(d3.behavior.zoom().on("zoom", function () {
                var translate = [d3.event.translate[0] + margin.left, d3.event.translate[1] + margin.top];
                svg.attr("transform", "translate(" + translate + ")" + " scale(" + d3.event.scale + ")")
            }))
            .attr("width", width + margin.right + margin.left)
            .attr("height", height + margin.top + margin.bottom)
            .append("g")
            .attr("transform", "translate(" + margin.left + "," + margin.top + ")");

    d3.select(self.frameElement).style("height", "800px");

    function getHaploGroupData(hgrp){
        d3.json("/gettree/"+hgrp, function(error, flare) {
            if (error) throw error;
            createTree(flare);

        });
    }




    function createTree(source){

        root = source;
        root.x0 = height / 2;
        root.y0 = 0;

        function collapse(d) {
            if (d.children) {
                d._children = d.children;
                d._children.forEach(collapse);
                d.children = null;
            }
        }

        root.children.forEach(collapse);
        update(root);
    }



    function update(source) {

        var spacing = 70;
        // Compute the new tree layout.
        var nodes = tree.nodes(root).reverse(),
                links = tree.links(nodes);

        // Normalize for fixed-depth.
        nodes.forEach(function(d) { d.y = d.depth * spacing; });

        // Update the nodes…
        var node = svg.selectAll("g.node")
                .data(nodes, function(d) { return d.id || (d.id = ++i); });

        // Enter any new nodes at the parent's previous position.
        var nodeEnter = node.enter().append("g")
                .attr("class", "node")
                .attr("transform", function(d) { return "translate(" + source.y0 + "," + source.x0 + ")"; })
                .on("click", click);

        nodeEnter.append("circle")
                .attr("r", 1e-6)
                .style("fill", function(d) { return d._children ? "red" : "#fff"; });

        nodeEnter.append("text")
                .attr("x", function(d) { return d.children || d._children ? 100 : -10; })
                .attr("dy", function(d){
                    console.log(d);
                    if(d.depth%2==0){
                        return "2em";
                    }else{
                        return "-1.35em";
                    }

                })
                .attr("text-anchor", function(d) { return d.children || d._children ? "end" : "start"; })
                .text(function(d) { return d.name; })
                .style("fill-opacity", 1e-6);

        // Transition nodes to their new position.
        var nodeUpdate = node.transition()
                .duration(duration)
                .attr("transform", function(d) { return "translate(" + d.y + "," + d.x + ")"; });

        nodeUpdate.select("circle")
                .attr("r", 10)
                .style("fill", function(d) { return d._children ? "red" : "#fff"; });

        nodeUpdate.select("text")
                .style("fill-opacity", 1);

        // Transition exiting nodes to the parent's new position.
        var nodeExit = node.exit().transition()
                .duration(duration)
                .attr("transform", function(d) { return "translate(" + source.y + "," + source.x + ")"; })
                .remove();

        nodeExit.select("circle")
                .attr("r", 1e-6);

        nodeExit.select("text")
                .style("fill-opacity", 1e-6);

        // Update the links…
        var link = svg.selectAll("path.link")
                .data(links, function(d) { return d.target.id; });

        // Enter any new links at the parent's previous position.
        link.enter().insert("path", "g")
                .attr("class", "link")
                .attr("d", function(d) {
                    var o = {x: source.x0, y: source.y0};
                    return diagonal({source: o, target: o});
                });

        // Transition links to their new position.
        link.transition()
                .duration(duration)
                .attr("d", diagonal);

        // Transition exiting nodes to the parent's new position.
        link.exit().transition()
                .duration(duration)
                .attr("d", function(d) {
                    var o = {x: source.x, y: source.y};
                    return diagonal({source: o, target: o});
                })
                .remove();

        // Stash the old positions for transition.
        nodes.forEach(function(d) {
            d.x0 = d.x;
            d.y0 = d.y;
        });
    }

    // Toggle children on click.
    function click(d) {
        if (d.children) {
            d._children = d.children;
            d.children = null;
        } else {
            d.children = d._children;
            d._children = null;
        }
        update(d);
    }


    function getHaplotypes(json){

    }

</script>


</body>


</html>